src/main.cpp

```diff
@@ -3906,15 +3940,23 @@ CBlockIndex* AddToBlockIndex(const CBlockHeader& block, const Consensus::Params&
         pindexNew->pprev = (*miPrev).second;
         pindexNew->nHeight = pindexNew->pprev->nHeight + 1;
 
-        if (IsActivationHeight(pindexNew->nHeight, consensusParams, Consensus::UPGRADE_HEARTWOOD)) {
-            // hashFinalSaplingRoot is currently null, and will be set correctly in ConnectBlock.
+        if (consensusParams.NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_NU5)) {
+            // The following hashes are currently null, and will be set correctly in
+            // ConnectBlock:
+            // - hashFinalSaplingRoot
+            // - hashFinalOrchardRoot
+            // - hashChainHistoryRoot
+        } else if (IsActivationHeight(pindexNew->nHeight, consensusParams, Consensus::UPGRADE_HEARTWOOD)) {
+            // hashFinalSaplingRoot and hashFinalOrchardRoot are currently null, and will
+            // be set correctly in ConnectBlock.
             // hashChainHistoryRoot is null.
         } else if (consensusParams.NetworkUpgradeActive(pindexNew->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
-            // hashFinalSaplingRoot is currently null, and will be set correctly in ConnectBlock.
-            pindexNew->hashChainHistoryRoot = pindexNew->hashLightClientRoot;
+            // hashFinalSaplingRoot and hashFinalOrchardRoot are currently null, and will
+            // be set correctly in ConnectBlock.
+            pindexNew->hashChainHistoryRoot = pindexNew->hashBlockCommitments;
         } else {
-            // hashChainHistoryRoot is null.
-            pindexNew->hashFinalSaplingRoot = pindexNew->hashLightClientRoot;
+            // hashFinalOrchardRoot and hashChainHistoryRoot are null.
```

Comment:
Can we assert this?

Reply:
Prior to Orchard activation, the value of `hashFinalOrchardRoot` doesn't really matter. `hashChainHistoryRoot` is constrained as previously noted.